From 5c01957e020a84814bb43230f532658469198f8c Mon Sep 17 00:00:00 2001
From: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Date: Tue, 1 Oct 2019 06:24:28 -0700
Subject: [PATCH 0677/1851] dma: xilinx: Disable low latency capture in file
 ops open

commit 56b1c8121b1de35a1ba3c80e2f914afbd3194c6d from
https://github.com/Xilinx/linux-xlnx.git

Add a custom handler to be called when file handle is opened
by application to disable the low latency capture and reset
dmaengine to AUTO_RESTART mode. We do this if the file handle
is the only file handle instance opened for the associated
video device.

This is required in the scenario where application got
hung or crash or is killed using SIGKILL and is not able to
disable the low latency capture by itself. After this scenario,
the next run was getting affected where even if low latency
capture was not set, driver was still running in low latency
mode due to previous context which was not cleared.

Signed-off-by: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Tested-by: Dylan Yip <dylan.yip@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 23 +++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 392ecd4cfc59..cfdc1034d2e0 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -1317,6 +1317,27 @@ static int xvip_dma_s_ctrl(struct v4l2_ctrl *ctl)
 	return ret;
 }
 
+static int xvip_dma_open(struct file *file)
+{
+	int ret;
+
+	ret = v4l2_fh_open(file);
+	if (ret)
+		return ret;
+
+	/* Disable the low latency mode as default */
+	if (v4l2_fh_is_singular_file(file)) {
+		struct xvip_dma *dma = video_drvdata(file);
+
+		mutex_lock(&dma->lock);
+		dma->low_latency_cap = false;
+		xilinx_xdma_set_mode(dma->dma, AUTO_RESTART);
+		mutex_unlock(&dma->lock);
+	}
+
+	return 0;
+}
+
 static const struct v4l2_ctrl_ops xvip_dma_ctrl_ops = {
 	.s_ctrl = xvip_dma_s_ctrl,
 };
@@ -1341,7 +1362,7 @@ static const struct v4l2_ctrl_config xvip_dma_ctrls[] = {
 static const struct v4l2_file_operations xvip_dma_fops = {
 	.owner		= THIS_MODULE,
 	.unlocked_ioctl	= video_ioctl2,
-	.open		= v4l2_fh_open,
+	.open		= xvip_dma_open,
 	.release	= vb2_fop_release,
 	.poll		= vb2_fop_poll,
 	.mmap		= vb2_fop_mmap,
-- 
2.31.1

